iceoryx_examples/singleprocess/single_process.cpp

```diff
@@ -44,38 +45,44 @@ void consoleOutput(const std::string& output)
 
 void sender()
 {
-    iox::popo::Publisher publisher({"Single", "Process", "Demo"});
+    iox::popo::UntypedPublisher publisher({"Single", "Process", "Demo"});
     publisher.offer();
 
     uint64_t counter{0};
     while (keepRunning.load())
     {
-        auto sample = static_cast<TransmissionData_t*>(publisher.allocateChunk(sizeof(TransmissionData_t)));
-        sample->counter = counter++;
-        consoleOutput(std::string("Sending: " + std::to_string(sample->counter)));
-        publisher.sendChunk(sample);
+        publisher.loan(sizeof(TransmissionData_t)).and_then([&](iox::popo::Sample<void>& sample) {
+            auto rawSample = static_cast<TransmissionData_t*>(sample.get());
+            rawSample->counter = counter++;
+            consoleOutput(std::string("Sending: " + std::to_string(rawSample->counter)));
+            sample.publish();
+        });
 
         std::this_thread::sleep_for(std::chrono::milliseconds(100));
     }
 }
 
 void receiver()
 {
-    iox::popo::Subscriber subscriber({"Single", "Process", "Demo"});
+    iox::popo::UntypedSubscriber subscriber({"Single", "Process", "Demo"});
 
     uint64_t cacheQueueSize = 10;
     subscriber.subscribe(cacheQueueSize);
 
     while (keepRunning.load())
     {
-        if (iox::popo::SubscriptionState::SUBSCRIBED == subscriber.getSubscriptionState())
+        if (iox::SubscribeState::SUBSCRIBED == subscriber.getSubscriptionState())
         {
             const void* rawSample = nullptr;
-            while (subscriber.getChunk(&rawSample))
+            while (subscriber.take().and_then([&](iox::cxx::optional<iox::popo::Sample<const void>>& maybeSample) {
```

Comment:
I think we have the same issue here that @elBoberido mentioned yesterday. Currently the wrong function is called in `expected::operator bool()`, see #351

Reply:
@ithier My first thought was to introduce a method `and_then_till_empty` (only valid with arg type `optional::T`) which loops as long as the `optional` has a value. Is that a good idea? Any other ideas how we could nicely loop in these kinds of situations? We use it quite often when dealing with chunks or queues.

Reply:
Another take: What about forwarding the `operator bool()` to the optional? This would change the semantic of the `expected` but one could still check for errors with `has_error()`.

Brief example:
````
template <typename ValueType, typename ErrorType>
template <typename Optional, typename std::enable_if<is_optional<Optional>::value, int>::type = 0>
expected<ValueType, ErrorType>::operator bool() const noexcept
{
    auto& nestedOptional = get_value();
    return nestedOptional.operator bool();
}
````

Reply:
> @ithier My first thought was to introduce a method `and_then_till_empty` (only valid with arg type `optional::T`) which loops as long as the `optional` has a value. Is that a good idea? Any other ideas how we could nicely loop in these kinds of situations? We use it quite often when dealing with chunks or queues.

@mossmaurice I think this is unsuitable for this problem. Here you assuming quiet much, namely that this is run in a loop which is often not the case. The `expected` should be a generic solution and specific solutions have to be developed by the user. 

Forwarding `operator bool()` to optional could be an interesting take on this. Maybe that we have a set of objects like:
- `optional`
- `expected`
- ?`shared_ptr`?
- ?`unique_ptr`
- ?`special_bool`? 

which provide the methods like:
- `and_then`
- `or_else`
- `operator bool()`

which are forwarded if they are combined in a template. This would make this really easy to use.
@ithier @elBoberido @FerdinandSpitzschnueffler @MatthiasKillat what do you think?

Reply:
I think we have to be very careful with this. Especially since the optional is already in the STL, people have an expectations how this works and when we break it, occasional contributors will get frustrated soon. The rants from a certain elf regarding an infamous STL implementation still resound in my head ;)
While this is neat for some cases, there might be corner cases where you really want to have the `operator bool()` operate on the `optional` and not on the wrapped object. I come more and more to the conclusion that we should make it visible to the developer that semantics changed, like calling `flatten()` before this behavior becomes active.

Reply:
I just realized that the std::optional doesn't have `and_then`/`or_else`, which meas we are free to define the standard, so forget what I said above :)

Reply:
@elBoberido nope we should somehow consider the C++ community and the ongoing discussions since `and_then` and `map` are already proposed and can become available in C++2x: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0798r0.html

Maybe we can get inspired a little bit by this article ...

Reply:
Here is the 3rd and most current revision of the `optional` proposal: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0798r3.html

Reply:
@elfenpiff do we want to adjust `cxx::optional` to that proposal

Reply:
@elBoberido  Absolutely! I created an issue for that: https://github.com/eclipse/iceoryx/issues/392

Reply:
@elfenpiff this might have to be done before v1.0 if there are breaking changes or wait for 2.0, though

Reply:
I would add the milestone Prio 3 here. If we have time we do it otherwise it is v2.0

Reply:
Not using the `bool()` operator now and added a separate `bool` to control the loop.